drivers/watchdog/wdt_sifive.c

```diff
@@ -0,0 +1,289 @@
+/*
+ * Copyright (C) 2020 Katsuhiro Suzuki
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @brief Watchdog (WDT) Driver for SiFive Freedom
+ */
+
+#define DT_DRV_COMPAT sifive_wdt
+
+#include <kernel.h>
+#include <soc.h>
+#include <drivers/watchdog.h>
+
+#define LOG_LEVEL CONFIG_WDT_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(wdt_sifive);
+
+#define WDOGCFG_SCALE_MAX     0xf
+#define WDOGCFG_SCALE_SHIFT   0
+#define WDOGCFG_SCALE_MASK    (WDOGCFG_SCALE_MAX << WDOGCFG_SCALE_SHIFT)
+#define WDOGCFG_RSTEN         BIT(8)
+#define WDOGCFG_ZEROCMP       BIT(9)
+#define WDOGCFG_ENALWAYS      BIT(12)
+#define WDOGCFG_COREAWAKE     BIT(13)
+#define WDOGCFG_IP0           BIT(28)
+
+#define WDOGCMP_MAX        0xffff
+
+#define WDOG_KEY           0x51f15e
+#define WDOG_FEED          0xd09f00d
+
+#define WDOG_CLK           32768
+
+struct wdt_sifive_reg {
+	/* offset: 0x000 */
+	uint32_t wdogcfg;
+	uint32_t dummy0;
+	uint32_t wdogcount;
+	uint32_t dummy1;
+	/* offset: 0x010 */
+	uint32_t wdogs;
+	uint32_t dummy2;
+	uint32_t wdogfeed;
+	uint32_t wdogkey;
+	/* offset: 0x020 */
+	uint32_t wdogcmp0;
+};
+
+struct wdt_sifive_device_config {
+	uintptr_t regs;
+};
+
+struct wdt_sifive_dev_data {
+	wdt_callback_t cb;
+	bool enable_cb;
+	bool timeout_valid;
+};
+
+#define DEV_CFG(dev) \
+	((const struct wdt_sifive_device_config *const)(dev)->config)
+#define DEV_REG(dev) \
+	((struct wdt_sifive_reg *)(DEV_CFG(dev))->regs)
+
+/**
+ * @brief Set maximum length of timeout to watchdog
+ *
+ * @param dev Watchdog device struct
+ */
+static void wdt_sifive_set_max_timeout(const struct device *dev)
+{
+	volatile struct wdt_sifive_reg *wdt = DEV_REG(dev);
+	uint32_t t;
+
+	t = wdt->wdogcfg;
+	t &= ~WDOGCFG_SCALE_MASK;
```

Comment:
Ah... I got a mistake. This line wants to set scale field 0xf:
t |= WDOGCFG_SCALE_MASK;

Reply:
Fixed